AST: Fix missing source loc in actor isolation diagnostics

AnthonyLatsis opened the pull request:
Form key path subscript components using the first valid location in the argument list instead of the '(' location, which is invalid in implicit argument lists containing explicit arguments. Such arguments lists are synthesized for the `@dynamicMemberLookup` feature.

AnthonyLatsis committed:
test: Merge `actor_keypath_isolation` & `actor_keypath_isolation_swift6`

AnthonyLatsis commented:
@swift-ci please smoke test macOS

xedin reviewed (approved):
I think it's the same situation as implicit vs. implicit + synthesized specially visible in derived synthesis related changes, we might want to consider splitting API into `forImplicitSingle` that always has a loc and `forSynthesizedSingle` that can default an an empty one.

xedin commented:
@swift-ci please test macOS platform

Xazax-hun reviewed (approved):
Importer change looks good to me.

hamishknight commented on lib/AST/ArgumentList.cpp:
```diff
@@ -132,9 +132,10 @@ ArgumentList::createImplicit(ASTContext &ctx, ArrayRef<Argument> args,
-ArgumentList *ArgumentList::forImplicitSingle(ASTContext &ctx, Identifier label,
-                                              Expr *arg) {
-  return createImplicit(ctx, {Argument(SourceLoc(), label, arg)});
+ArgumentList *ArgumentList::forImplicitSingle(ASTContext &ctx, SourceLoc loc,
+                                              Identifier label, Expr *arg) {
+  return createImplicit(ctx, /*lParenLoc=*/loc, {Argument(loc, label, arg)},
+                        /*lParenLoc=*/loc);
```
One potential concern with this is that it could result in out-of-order source ranges where the arg expression extends past the argument list, IMO it would make more sense to compute the paren locs from the start and end location of the expression.

An alternative fix here would be to change `KeyPathExpr::Component::for(Unresolved)Subscript` to use `ArgumentList::getStartLoc` instead of `ArgumentList::getLParenLoc`, I don't feel too strongly either way though.

AnthonyLatsis replied:
Why is this a concern for implicit expressions? Like hitting an assert in source range computation where the start is past the end?

hamishknight replied:
It's just a useful invariant that we ought to be able to enforce for the AST in general, though currently the ASTVerifier avoids checking implicit nodes

AnthonyLatsis replied:
> An alternative fix here would be to change `KeyPathExpr::Component::for(Unresolved)Subscript` to use `ArgumentList::getStartLoc` instead of `ArgumentList::getLParenLoc`

This makes more sense to me, thank you!

AnthonyLatsis commented:
> I think it's the same situation as implicit vs. implicit + synthesized

Do we actually have a precedent for implicit code that never needs a location?

xedin commented:
> Do we actually have a precedent for implicit code that never needs a location?

I think that's mostly implicit + synthesized case.

AnthonyLatsis committed:
AST: Fix missing source loc in actor isolation diagnostics

Form key path subscript components using the first valid location in
the argument list instead of the '(' location, which is invalid in
implicit argument lists containing explicit arguments. Such arguments
lists are synthesized for the `@dynamicMemberLookup` feature.

AnthonyLatsis committed:
Concurrency: Add note to diagnostic about isolated key path component

...when the component is a synthesized reference to a key-path-based
`@dynamicMemberLookup` subscript, for clarify.

AnthonyLatsis commented:
@swift-ci please smoke test

AnthonyLatsis commented:
@swift-ci please smoke test

hamishknight commented on lib/Sema/TypeCheckConcurrency.cpp:
```diff
@@ -4197,6 +4197,15 @@ namespace {
               .warnUntilSwiftVersionIf(downgrade, 6);
 
+            // If the culprit is a synthesized reference to a
+            // `@dynamicMemberLookup` subscript, point to it for clarify.
```
```suggestion
            // `@dynamicMemberLookup` subscript, point to it for clarity.
```
